docs(cn): translate reference/react/cloneElement into Chinese (#1207)
reactjs/zh-hans.react.dev: src/content/reference/react/cloneElement.md

console.log(clonedElement); // <Row title="Cabbage">Goodbye</Row>
```

> [请参阅下面的更多示例](#usage)

Reviewer: 标点缺失。

Author: done~

#### 参数 {/*parameters*/}

> * `element`：`element` 参数必须是一个有效的 React 元素。例如它可以是一个像 `<Something />` 这样的 JSX 节点，调用 [`createElement`](/reference/react/createElement) 的结果，或者另一个 `cloneElement` 调用的结果。

Reviewer: 后半句略有拗口。

Author: 优化了一下

> * `props`：`props` 参数必须是一个对象或 `null`。 如果传 `null`，克隆后的元素将保留所有原始的 `element.props`。 否则，对于 `props` 对象中的每个 prop 属性，返回的元素将“优先”使用 `props` 中的值而不是 `element.props` 中的值。 其余的 props 将从原始的 `element.props` 中填充。如果你传递 `props.key` 或 `props.ref`，它们将替换原来的。

Reviewer 2: 汉字和中文标点之间不应有空格。请调整。

Author: done~

* **可选** `...children`：零个或多个子节点。它们可以是任何 React 节点，包括 React 元素、字符串、数字、[portals](/reference/react-dom/createPortal)、空节点（`null`、`undefined`、`true` 和 `false`），和 React 元素数组。如果你不传递任何 `...children` 参数，则原始的 `element.props.children` 将被保留。

#### 返回值 {/*returns*/}

`cloneElement` 返回一个具有一些属性的 React element 对象：

* `type`：与 `element.type` 相同。
* `props`：将 `element.props` 与你传递的 `props` 浅合并的结果。
* `ref`：原始的 `element.ref`，除非它被 `props.ref` 覆盖。
* `key`：原始的 `element.key`，除非它被 `props.key` 覆盖。

通常，你将从组件返回该元素或使其成为另一个元素的子元素。尽管你可以读取元素的属性，但最好在创建每个元素后将其视为不透明的，并且仅渲染它。

#### 注意事项 {/*caveats*/}

* 克隆一个元素 **不会修改原始元素**。

> * 如果 children 已知是静态的，则你应该 **将它们作为多个参数传递给 `cloneElement`**，例如 `cloneElement(element, null, child1, child2, child3)`。如果你的 children 是动态的，请将整个数组作为第三个参数传递：`cloneElement(element, null, listItems)`。这确保了 React 会对任何动态列表 [警告你缺少“key”](/learn/rendering-lists#keeping-list-items-in-order-with-key)，对于静态的列表，这是不必要的，因为它们不会重新排序。

Reviewer: ```suggestion
* 如果已知 children 是静态的，则你应该 **将它们作为多个参数传递给 `cloneElement`**，例如 `cloneElement(element, null, child1, child2, child3)`。如果你的 children 是动态的，请将整个数组作为第三个参数传递：`cloneElement(element, null, listItems)`。这确保了 React 会对任何动态列表 [警告你缺少“key”](/learn/rendering-lists#keeping-list-items-in-order-with-key)，对于静态的列表，这是不必要的，因为它们不会重新排序。
```

);
```

在这里，生成的 <CodeStep step={3}>克隆 element</CodeStep> 将为 `<Row title="Cabbage" isHighlighted={true} />`。

> **让我们看一个示例，看看它什么时候有用。**

Reviewer: ```suggestion
**让我们看一个示例，看看它什么时候有用**。
```

> 想象一个 `List` 组件将其 [`children`](/learn/passing-props-to-a-component#passing-jsx-as-children) 呈现为可选择行的列表，并带有可更改的“下一步”按钮选择了哪一行。`List` 组件需要以不同的方式渲染所选的 `Row`，因此它克隆它收到的每个 `<Row>` 子级，并添加额外的 `isHighlighted: true` 或 `isHighlighted: false` 属性：

Reviewer: ```suggestion
想象一个 `List` 组件将其 [`children`](/learn/passing-props-to-a-component#passing-jsx-as-children) 渲染为可选择行的列表，并带有可更改的“下一步”按钮选择了哪一行。`List` 组件需要以不同的方式渲染所选的 `Row`，因此它克隆它收到的每个 `<Row>` 子级，并添加额外的 `isHighlighted: true` 或 `isHighlighted: false` 属性：
```

</Sandpack>

> 总而言之，`List` 克隆了它接收的 `<Row />` 元素，并向它们添加了一个额外的 prop。

Reviewer: ```suggestion
总而言之，`List` 克隆了它接收的 `<Row />` 元素，并向它们添加额外的 props。
```

<Pitfall>

克隆 children 使得你很难判断数据如何流经你的应用。尝试一种 [替代方案](#alternatives)。

</Pitfall>

---

## 替代方案 {/*alternatives*/}

### 通过 props 传递数据 {/*passing-data-with-a-render-prop*/}

> 接受类似 `renderItem` 这样的 *render prop* 代替 `cloneElement` 的用法。在这里，`List` 接收 `renderItem` 作为 prop。`List` 为数组每一项调用 `renderItem`，并传递 `isHighlighted` 作为参数：

Reviewer: ```suggestion
接受类似 `renderItem` 这样的 *render prop* 代替 `cloneElement` 的用法。在这里，`List` 接收 `renderItem` 作为 props。`List` 为数组每一项调用 `renderItem`，并传递 `isHighlighted` 作为参数：
```

Reviewer 3: 为啥原文 `as a prop` 能翻译成 `props`

Reviewer: 在翻译为中文时，如果是表达 React props 这样整体的概念的时候，优先使用 props；其次才考虑原文的单数 prop 是否是特指到某一个具体的 prop，然后保留单数 prop。
此处原文 as a prop 尽管是使用的单数，但是表达的意思更加倾向于前者所述的整体的概念，即是作为组件的 props（尽管只有一个），而不是作为 props 的“某一个 prop”，所以考虑使用 props。

</Sandpack>

这种方案优于 `cloneElement`，因为它更加清晰。

---

### 通过 context 传递数据 {/*passing-data-through-context*/}

`cloneElement` 的另一种替代方法是 [通过 context 传递数据](/learn/passing-data-deeply-with-context)。

例如，你可以调用 [`createContext`](/reference/react/createContext) 来定义一个 `HighlightContext`：

```js
export const HighlightContext = createContext(false);
```

> 你的 `List` 组件可以将其呈现的每个 item 传递到 `HighlightContext` provider 中：

Reviewer: ```suggestion
`List` 组件可以将其呈现的每个 item 传递到 `HighlightContext` provider 中：
```

      })}
```

> 通过这种方法，`Row` 不需要接收 `isHighlighted`属性，它可以从上下文读取到：

Reviewer: ```suggestion
通过这种方法，`Row` 不需要接收 `isHighlighted`属性，因为它可以从 context 中读取：
```
